bitcoin/script/miniscript.h

```diff
@@ -348,6 +349,107 @@ struct Node {
+     */
+    template<typename Result, typename State, typename DownFn, typename UpFn>
+    std::optional<Result> TreeEvalMaybe(State root_state, DownFn downfn, UpFn upfn) const
+    {
+        /** Entries of the explicit stack tracked in this algorithm. */
+        struct StackElem
+        {
+            const Node& node; //!< The node being evaluated.
+            bool expanded; //!< Whether the children of this node have been added.
+            State state; //!< The state for that node.
+
+            StackElem(const Node& node_, bool exp_, State&& state_) :
+                node(node_), expanded(exp_), state(std::move(state_)) {}
+        };
+        /* Stack of tree nodes being explored. */
+        std::vector<StackElem> stack;
+        /* Results of subtrees so far. Their order and mapping to tree nodes
+         * is implicitly defined by stack. */
+        std::vector<Result> results;
+        stack.emplace_back(*this, false, std::move(root_state));
+
+        /* Here is a demonstration of the algorithm, for an example tree A(B,C(D,E),F).
+         * State variables are omitted for simplicity.
+         *
+         * First: stack=[(A,false)] results=[].
+         *        stack=[(A,true),(F,false),(C,false),(B,false)] results=[]
+         *        stack=[(A,true),(F,false),(C,false)] results=[B]
+         *        stack=[(A,true),(F,false),(C,true),(E,false),(D,false)] results=[B]
+         *        stack=[(A,true),(F,false),(C,true),(E,false)] results=[B,D]
+         *        stack=[(A,true),(F,false),(C,true)] results=[B,D,E]
+         *        stack=[(A,true),(F,false)] results=[B,C]
+         *        stack=[(A,true)] results=[B,C,F]
+         * Final: stack=[] results=[A]
+         */
+        while (stack.size()) {
+            const Node& node = stack.back().node;
+            if (!stack.back().expanded && !node.subs.empty()) {
+                /* We encounter an unexpanded tree node with children. Mark it as
+                 * expanded, and add its children in reverse order. By the time we
+                 * process this node again, the results of the children will be
+                 * at the end of results in normal order. */
+                stack.back().expanded = true;
+                stack.reserve(stack.size() + node.subs.size());
+                const auto& parent_state = stack.back().state;
+                for (size_t i = 0; i < node.subs.size(); ++i) {
+                    size_t index = node.subs.size() - i - 1;
+                    stack.emplace_back(**(node.subs.rbegin() + i), false,
+                        downfn(parent_state, node, index));
+                }
+                continue;
+            }
+            // Invoke upfn with the last node.subs.size() elements of results as input.
+            assert(results.size() >= node.subs.size());
+            std::optional<Result> result{upfn(std::move(stack.back().state), node,
+                Span{results}.last(node.subs.size()))};
```

Comment:
```suggestion
                Span<Result>{results}.last(node.subs.size()))};
```

Reply:
That really shouldn't be necessary. Do you have an old span.h by any chance, or not using -std=c++17?

Reply:
Interestingly this only happened when building with fuzz enabled. I'll retry both builds from scratch and report.

Reply:
So i still have this error (only) when building the fuzz target.
```
$ make V=1
Making all in src
make[1]: Entering directory '/home/darosior/projects/bitcoin/src'
make[2]: Entering directory '/home/darosior/projects/bitcoin/src'
make[3]: Entering directory '/home/darosior/projects/bitcoin'
make[3]: Leaving directory '/home/darosior/projects/bitcoin'
clang++ -std=c++17 -DHAVE_CONFIG_H -I. -I../src/config  -DABORT_ON_FAILED_ASSUME -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=2 -I. -I./minisketch/include -I./secp256k1/include -I./univalue/include -I/home/darosior/projects/bitcoin/../db4/include -I/usr/include -I./leveldb/include -I./leveldb/helpers/memenv -DHAVE_BUILD_INFO  -Wstack-protector -fstack-protector-all -fcf-protection=full -fstack-clash-protection -Wall -Wextra -Wgnu -Wformat -Wformat-security -Wvla -Wshadow-field -Wthread-safety -Wrange-loop-analysis -Wredundant-decls -Wunused-member-function -Wdate-time -Wconditional-uninitialized -Woverloaded-virtual -Wsuggest-override -Wunreachable-code-loop-increment -Wimplicit-fallthrough -Wno-unused-parameter -Wno-self-assign -Wno-deprecated-copy   -fsanitize=fuzzer  -fPIE -g -O2 -MT test/fuzz/fuzz-miniscript_decode.o -MD -MP -MF test/fuzz/.deps/fuzz-miniscript_decode.Tpo -c -o test/fuzz/fuzz-miniscript_decode.o `test -f 'test/fuzz/miniscript_decode.cpp' || echo './'`test/fuzz/miniscript_decode.cpp
In file included from test/fuzz/miniscript_decode.cpp:8:
./script/miniscript.h:425:30: error: member reference base type 'Span' is not a structure or union
                Span{results}.last(node.subs.size()))};
                ~~~~~~~~~~~~~^~~~~
1 error generated.
make[2]: *** [Makefile:15058: test/fuzz/fuzz-miniscript_decode.o] Error 1
make[2]: Leaving directory '/home/darosior/projects/bitcoin/src'
make[1]: *** [Makefile:17529: all-recursive] Error 1
make[1]: Leaving directory '/home/darosior/projects/bitcoin/src'
make: *** [Makefile:823: all-recursive] Error 1
```
```
$ clang++ --version
Debian clang version 11.0.1-2
```

Reply:
@darosior Where is the code for that?

Reply:
Here you go: https://github.com/darosior/bitcoin/tree/miniscript_pr_83. Based on the watchonly PR (https://github.com/darosior/bitcoin/pull/5).

I just re-tried on this branch tip and i can build without issue with:
```shell
./autogen.sh
./configure --enable-debug --enable-c++17 --without-gui --disable-bench "$@" BDB_LIBS="-L${BDB_PREFIX}/lib -ldb_cxx-4.8" BDB_CFLAGS="-I${BDB_PREFIX}/include"
make -j20
```

But not with
```shell
./autogen.sh
./configure --enable-fuzz --enable-c++17 --with-sanitizers=fuzzer BDB_LIBS="-L${BDB_PREFIX}/lib -ldb_cxx-4.8" BDB_CFLAGS="-I${BDB_PREFIX}/include" CC=clang CXX=clang++
make -j20
```

And the `g++` used to build without the fuzz target:
```
g++ (Debian 10.3.0-11) 10.3.0
```

Reply:
Very strange. I can build the fuzz tests without problem in your branch (clang++ --version => 12.0.0-3ubuntu1~21.04.2). I'll add the `Span<Result>` if that helps, but I'd still like to understand what causes this.

Reply:
Trying with another `clang` version now.